c-deps/libroach/incremental_iterator.cc

```diff
@@ -95,20 +51,106 @@ bool DBIncrementalIterator::legacyTimestampIsLess(const cockroach::util::hlc::Le
          (t1.wall_time() == t2.wall_time() && t1.logical() < t2.logical());
 }
 
-// advanceKey finds the key and its appropriate version which lies in
+// extractKey extracts the key portion of the mvcc_key and put it in key. It
+// returns a validity indicator.
+WARN_UNUSED_RESULT bool DBIncrementalIterator::extractKey(rocksdb::Slice mvcc_key, rocksdb::Slice *key) {
+  rocksdb::Slice ts;
+  if (!SplitKey(mvcc_key, key, &ts)) {
+    valid = false;
+    status = FmtStatus("failed to split mvcc key");
+    return false;
+  }
+  return true;
+}
+
+// maybeSkipKeys checks if any keys can be skipped by using a time-bound
+// iterator. If keys can be skipped, it will update the main iterator to point
+// to the earliest version of the next candidate key.
+// It is expected that TBI is at a key <= main iterator key when calling
+// maybeSkipKeys().
+void DBIncrementalIterator::maybeSkipKeys() {
+  if (time_bound_iter == nullptr) {
+    // We don't have a TBI, so we can't skip any keys.
+    return;
+  }
+
+  rocksdb::Slice tbi_key;
+  if(!extractKey(time_bound_iter->rep->key(), &tbi_key)) {
+    return;
+  }
+  rocksdb::Slice iter_key;
+  if(!extractKey(iter_key->rep->key(), &iter_key)) {
+    return;
+  }
+
+  if (iter_key.compare(tbi_key) > 0) {
+    // If the iterKey got ahead of the TBI key, advance the TBI Key.
+    auto state = DBIterNext(time_bound_iter.get(), true /* skip_current_key_versions */);
+    if (!state.valid) {
+        status = state.status;
+        valid = false;
+        return;
+    }
+    if(!extractKey(time_bound_iter->rep->key(), &tbi_key)) {
+      return;
+    }
+
+    auto cmp = iter_key.compare(tbi_key);
+    // The case where iterKey.Compare(tbiKey) == 0 is the fast-path is when the
+    // TBI and the main iterator are in lockstep. In this case, the main
+    // iterator was referencing the next key that would be visited by the TBI.
+    // This means that for the incremental iterator to perform a Next or NextKey
+    // will require only 1 extra NextKey invocation while they remain in
+    // lockstep.
+
+    if (cmp > 0) {
+      // If the tbiKey is still behind the iterKey, the TBI key may be seeing
+      // phantom MVCCKey.Keys. These keys may not be seen by the main iterator
+      // due to aborted transactions and keys which have been subsumed due to
+      // range tombstones. In this case we can SeekGE() the TBI to the main iterator.
+
+      // NB: Seek() is expensive, and this is only acceptable since we expect
+      // this case to rarely occur.
```

Comment:
I feel like this important note belongs right at the beginning of this if statement -- it explains why we call Next, as opposed to seeking right away.